casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -422,4 +426,77 @@ object ProtoUtil {
     (missingParents union missingJustifications).toList
   }
 
+  // Return hashes of all blocks that are yet to be seen by the passed in block
+  def unseenBlockHashes[F[_]: Sync: BlockStore](
+      dag: BlockDagRepresentation[F],
+      block: BlockMessage
+  ): F[Set[BlockHash]] =
+    for {
+      latestMessages        <- dag.latestMessages
+      latestMessagesOfBlock <- ProtoUtil.toLatestMessage(block.justifications, dag)
+      result <- latestMessages.toList
+                 .traverse {
+                   case (validator, latestMessage) =>
+                     latestMessagesOfBlock.get(validator) match {
+                       case Some(latestMessageOfBlockByValidator) =>
+                         DagOperations
+                           .bfTraverseF(List(latestMessage))(
+                             block =>
+                               getCreatorJustificationUnlessGoal(
+                                 dag,
+                                 block,
+                                 latestMessageOfBlockByValidator
+                               )
+                           )
+                           .map(_.blockHash)
+                           .toSet
+                       case None =>
+                         Set.empty[BlockHash].pure[F]
+                     }
+                 }
+                 .map(_.flatten.toSet)
+    } yield result -- latestMessagesOfBlock.values.map(_.blockHash) - block.blockHash
```

Comment:
The body of the traverse, being so big, deserves a name and a type signature in order to be easier to understand. I find something like the code below to be more readable if you replace `computeSomething` and `result` with proper names.

```scala
  def unseenBlockHashes[F[_]: Sync: BlockStore](
      dag: BlockDagRepresentation[F],
      block: BlockMessage
  ): F[Set[BlockHash]] = {
    def computeSomething(
        latestMessagesOfBlock : Map[Validator, BlockMetadata], latestMessages : Map[Validator, BlockMetadata]) (
        validator : Validator, latestMessage : BlockMetadata
    ) : F[Set[ByteString]] =
      latestMessagesOfBlock
        .get(validator)
        .map(latestMessageOfBlockByValidator =>
          DagOperations
            .bfTraverseF(List(latestMessage))(
              block =>
                getCreatorJustificationUnlessGoal(
                  dag,
                  block,
                  latestMessageOfBlockByValidator
                )
            )
            .map(_.blockHash)
            .toSet
        )
        .getOrElse(Set.empty[BlockHash])

    for {
      latestMessages <- dag.latestMessages
      latestMessagesOfBlock <- ProtoUtil.toLatestMessage(block.justifications, dag)
      result <- 
        latestMessages
          .toStream
          .traverse(Function.tupled(computeSomething(latestMessagesOfBlock, latestMessages)))
    } yield result.flatten.toSet -- latestMessagesOfBlock.values.map(_.blockHash) - block.blockHash
  }
```